relayer/src/chain/handle.rs

```diff
@@ -365,9 +367,10 @@ pub trait ChainHandle: Clone + Send + Sync + Serialize + Debug {
 
     fn get_key(&self) -> Result<KeyEntry, Error>;
 
-    fn add_key(&self, key_name: String, key: KeyEntry) -> Result<(), Error>;
+    fn app_version(&self, request: QueryAppVersionRequest)
+        -> Result<ics04_channel::Version, Error>;
```

Comment:
Let's rather expose the underlying fields (as domain types) rather than the proto type itself both because we do not want to depend on proto types in public APIs but also because non-Cosmos chains may use different RPC mechanism than gRPC.

```suggestion
    fn app_version(
        &self,
        port_id: PortId,
        connection_id: ConnectionId,
        ordering: Odering,
        counterparty: Option<Counterparty>,
        proposed_version: Version,
    ) -> Result<ics04_channel::Version, Error>;
```

Reply:
Hmm I thought about this a lot and I'm hesitant to do the change. I'll describe below why I'm hesitating, but would encourage that in we take your points into account, Romain, when we work on https://github.com/informalsystems/ibc-rs/discussions/1318, #1487, and #1001.

**Reason to change to domain types:** Briefly, for clean APIs and better modularity wrt supporting different chain types. 
> both because we do not want to depend on proto types in public APIs but also because non-Cosmos chains may use different RPC mechanism than gRPC.

**Reason to keep the proto types:** Briefly, for developer experience.

This `QueryAppVersionRequest` appearing in our public API is an IBC.core type. Any IBC-enabled chain will support this query. (It can be implemented differently than gRPC, but not sure that's a good reason to avoid using the `QueryAppVersionRequest` data type).

But `QueryAppVersionRequest` being an IBC core is not the the point. The point is: We want to have simple interfaces in `ChainHandle` trait because of the handle <> runtime architecture. If the `app_version` method will take five parameters, all those parameters have to be carried back and forth through ~8 interface (or maybe more): chain handle, ProdChainHandle, integration test chain handle, runtime twice, `ChainRequest` enum, cosmos.rs and mock.rs impl. This is too much. So I think the methods signatures should be very compact, though the API can suffer as a consequence as it does in this instance (it doesn't have to, though, we can have both clean APIs and good developer experience, but that's maybe for future PR).

Let me know if I missed anything.

Reply:
> It can be implemented differently than gRPC, but not sure that's a good reason to avoid using the QueryAppVersionRequest data type

imho that's the first reason for not exposing the proto type, as it might be Cosmos-specific and may not be relevant for other types of chains.

Additionally, the idea with having both proto and domain types in tendermint-rs and ibc-rs is that the proto types are only relevant to (de)serialization and are otherwise an implementation detail which should not leak in public APIs.

What I would suggest we do here if you are worried about having too many parameters, is to define our own `QueryAppVersion` struct which would have those params as fields and then pass just this single struct around.

Does that make sense?

Reply:
Thanks for following-up Romain. I agree with your suggestion, done! dc3e0c2ed6c836c77ebd04c7be90ea3e36703ee0